scipy/optimize/_differentialevolution.py

```diff
@@ -861,18 +1045,37 @@ def __next__(self):
                 parameters = self._scale_parameters(trial)
 
                 # determine the energy of the objective function
-                energy = self.func(parameters)
-                self._nfev += 1
-
-                # if the energy of the trial candidate is lower than the
-                # original population member then replace it
-                if energy < self.population_energies[candidate]:
+                if self.constraints:
+                    cv = self._constraint_violation_fn(parameters)
+                    feasible = False
+                    energy = np.inf
+                    if not np.sum(cv) > 0:
+                        # solution is feasible
+                        feasible = True
+                        energy = self.func(parameters)
+                        self._nfev += 1
+                else:
+                    feasible = True
+                    cv = np.atleast_2d([0.])
+                    energy = self.func(parameters)
+                    self._nfev += 1
+
+                # compare trial and population member
+                if self._accept_trial(energy, feasible, cv,
+                                      self.population_energies[candidate],
+                                      self.feasible[candidate],
+                                      self.constraint_violation[candidate]):
                     self.population[candidate] = trial
                     self.population_energies[candidate] = energy
-
-                    # if the trial candidate also has a lower energy than the
-                    # best solution then promote it to the best solution.
-                    if energy < self.population_energies[0]:
+                    self.feasible[candidate] = feasible
+                    self.constraint_violation[candidate] = cv
+
+                    # if the trial candidate is also better than the best
+                    # solution then promote it.
+                    if self._accept_trial(energy, feasible, cv,
```

Comment:
Based on Lampinen alone, I'm not sure what promoting is all about. Lampinen specifically writes: "The interesting point concerning DE's replacement scheme is that a trial vector is only compared to one individual, not to all the individuals in the current population." Maybe this is why you suggested using the "deferred" option? I'll have to take a more careful look at how promotion works.

Reply:
Refer to equation 5. In the rand1bin strategy `r1, r2, r3` are randomly selected. In the best1bin strategy `r3` is not randomly selected, it's the best solution so far. 

In this implementation the best solution is kept in the first population entry. The `_promote_lowest_energy` method finds the best solution so far and moves it into the first population entry, with the previous best solution moving into its place.

If `updating=immediate` then if the trial vector is better than the best solution so far then it's immediately promoted to the best solution so subsequent trial vectors within the same generation can take advantage of the best solution.
If `updating=deferred` then the best solution is only updated once per generation.